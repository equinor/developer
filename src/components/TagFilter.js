import { useState } from "react";
import styled from "styled-components";
import React from "react";
import { ListItem, TagWrapper } from "./TagListing";

/**
 * @param selectedTags InitialSelectedTags
 * @returns {function({node: *}): boolean}
 */
export function filterTags(selectedTags) {
  const hasSelections = selectedTags.filter(tag => tag.selected).length;
  return ({ node }) => {
    if (!hasSelections || !node.frontmatter.tags) {
      return true;
    }

    //returns true if one of post tag's is selected.
    const tags = node.frontmatter.tags.filter(tagName => {
      return selectedTags.filter(selectedTag => {
        return selectedTag.selected && selectedTag.name === tagName;
      }).length;
    });
    return tags.length;
  };
}

export function initializeSelectedTags(tags, paramTag) {
  return tags.map(tag => {
    const selected = paramTag ? paramTag === tag.fieldValue : true;
    return { name: tag.fieldValue, selected };
  });
}

const FilterListItem = styled(ListItem)`
  cursor: pointer;
`;

/**
 * SelectedTags should be available outside this component.
 * Passing selectedTags to children will make TagFilter less usable.
 *
 * @param tags array {fieldValue}
 * @param selectedTags array {name: string, selected: boolean}
 * @param setSelectedTags setState(selectedState)
 * @returns {*} component
 * @constructor
 */
export const TagFilter = ({ tags, selectedTags, setSelectedTags }) => {
  const [selectAll, setSelectAll] = useState(false);

  const TagStatus = styled.span`
    font-size: 24px;
    padding-left: 10px;
    fontweight: 500;
  `;

  const toggleTags = () => {
    setSelectAll(!selectAll);
    setSelectedTags(
      selectedTags.map(tag => {
        tag.selected = selectAll;
        return tag;
      })
    );
  };

  const toggleTag = (tags, name) => {
    setSelectedTags(
      tags.map(tag => {
        //toggle single tag if name is provided.
        if (tag.name === name) {
          tag.selected = !tag.selected;
        }
        return tag;
      })
    );
  };

  const SelectAll = ({ selectAll, onClick }) => (
    <span
      style={{ textDecoration: "underline", cursor: "pointer" }}
      onClick={onClick}
    >
      {selectAll ? "Select all" : "Deselect all"}
    </span>
  );
  return (
    <TagWrapper>
      {tags.map((tag, index) => {
        const isEnabled = selectedTags[index].selected;
        return (
          <FilterListItem
            enabled={isEnabled}
            key={tag.fieldValue}
            onClick={() => toggleTag(selectedTags, tag.fieldValue)}
          >
            <span>
              {tag.fieldValue}
              <TagStatus>{isEnabled ? "X" : "+"}</TagStatus>
            </span>
          </FilterListItem>
        );
      })}
      <SelectAll onClick={toggleTags} selectAll={selectAll} />
    </TagWrapper>
  );
};
