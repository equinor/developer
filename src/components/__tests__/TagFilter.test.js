import renderer from "react-test-renderer";
import React from "react";
import { render, fireEvent } from "react-testing-library";
import { TagFilter } from "../TagFilter";

const tags = [{ fieldValue: "Api" }, { fieldValue: "Tech", selected: true }];

//generate test data
const selectedTags = tags.map(tag => ({
  name: tag.fieldValue,
  selected: tag.selected,
}));

describe("TagFilter", () => {
  it("renders correctly", () => {
    const tree = renderer
      .create(<TagFilter tags={tags} selectedTags={selectedTags} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should deselect all tags", () => {
    const mockClick = jest.fn();
    const { getByText } = render(
      <TagFilter
        tags={tags}
        selectedTags={selectedTags}
        setSelectedTags={mockClick}
      />
    );
    fireEvent(
      getByText("Deselect all"),
      new MouseEvent("click", {
        bubbles: true,
        cancelable: true,
      })
    );
    expect(mockClick).toHaveBeenCalled();
  });

  it("should toggle tag", () => {
    const mockClick = jest.fn();
    const { getByText } = render(
      <TagFilter
        tags={tags}
        selectedTags={selectedTags}
        setSelectedTags={mockClick}
      />
    );
    fireEvent(
      getByText("Tech"),
      new MouseEvent("click", {
        bubbles: true,
        cancelable: true,
      })
    );
    expect(mockClick).toHaveBeenCalled();
  });
});
