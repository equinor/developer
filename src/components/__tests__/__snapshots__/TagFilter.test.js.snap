// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagFilter renders correctly 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-left: 15px;
  margin-right: 15px;
  margin: 0 15px 40px;
}

.c2 {
  margin-top: 10px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  font-size: 16px;
  font-weight: bold;
  line-height: 21px;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  margin-right: 15px;
}

.c1 {
  margin-top: 40px;
  list-style: none;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  padding-left: 20px;
}

.c3 {
  background-color: #EAEAEA;
  padding: 0 20px;
  text-align: center;
  border-radius: 20.5px;
  margin-right: 20px;
  margin-bottom: 10px;
  cursor: pointer;
}

.c3 a,
.c3 span {
  font-size: 12px;
  font-weight: 500;
  line-height: 14px;
  color: #007079;
  text-transform: uppercase;
}

.c5 {
  background-color: #DEFAEB;
  padding: 0 20px;
  text-align: center;
  border-radius: 20.5px;
  margin-right: 20px;
  margin-bottom: 10px;
  cursor: pointer;
}

.c5 a,
.c5 span {
  font-size: 12px;
  font-weight: 500;
  line-height: 14px;
  color: #007079;
  text-transform: uppercase;
}

.c4 {
  font-size: 24px;
  padding-left: 10px;
  fontweight: 500;
}

<section
  className="c0"
>
  <ul
    className="c1"
  >
    <span
      className="c2"
    >
      Tags
    </span>
    <li
      className="c3"
      onClick={[Function]}
    >
      <span>
        Api
        <span
          className="c4"
        >
          +
        </span>
      </span>
    </li>
    <li
      className="c5"
      onClick={[Function]}
    >
      <span>
        Tech
        <span
          className="c4"
        >
          X
        </span>
      </span>
    </li>
    <span
      onClick={[Function]}
      style={
        Object {
          "cursor": "pointer",
          "textDecoration": "underline",
        }
      }
    >
      Deselect all
    </span>
  </ul>
</section>
`;
